src/bin_pattern.jl

```diff
@@ -0,0 +1,169 @@
+"""
+    bin_sparse_matrix!(M::AbstractArray, M_id::AbstractSparseMatrixCSC, max_num_bins::Integer)
+
+Compute a Binning pattern in `M_id` from the sparsity pattern in `M_id`.
+
+`M_id` is SparseMatrixCSC{Int64, Int64} with the same shape as `M`,
+when handed over it contains only 0 and 1. After usage it contains an Integer
+corresponding to a bin or 0. According to the Binning Pattern,
+similar valued entries can be considered the same.
+
+`M` is the matrix whose entries are used for binning. `max_num_bins` is the maximum
+number of bins and must be non-negative, if `max_num_bins=0`, no binning is
+performed and only the sparsity pattern in `M_id` is used then.
+
+See also: [`sps_compute`](@ref), [`p_norm_sparsity_matrix`](@ref).
+
+# Examples
+```jldoctest
+julia> bin_sparse_matrix!([4 1 4.01; 0.1 17.1 17; 0.2 4 29], sparse([1 0 1; 0 1 1; 0 0 1]), 200)
+3×3 SparseMatrixCSC{Int64, Int64} with 5 stored entries:
+ 1  ⋅  1
+ ⋅  2  2
+ ⋅  ⋅  3
+```
+"""
+function bin_sparse_matrix!(M::AbstractArray{T}, M_id::AbstractSparseMatrixCSC, max_num_bins::Integer) where {T}
+    @assert max_num_bins >= 0 "max_num_bins must be non-negative"
+    @assert size(M, 1) == size(M_id, 1) && size(M, 2) == size(M_id, 2) "M must be the same size as M_id"
+
+    if T <: Integer
+        M = 1.0*M
+    end
+    P = eltype(M)
```

Comment:
This might be bad for type stability and therefore performance.

What you could do is to have an extra function
```julia
function bin_sparse_matrix!(
    M::AbstractArray{<:Int}, 
    M_id::AbstractSparseMatrixCSC, 
    max_num_bins::Integer,
)
    return bin_sparse_matrix!(float(M), M_id, max_num_bins)
end
```
This makes use of multiple dispatch and will only be used on integer arrays.

Reply:
thats nice! but didnt my code do the same? if M<:Int it makes it to a float matrix, else it uses the type of M right?